cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,973 @@
+a `string_scalar` holds a single string. The data for the stored value resides in device memory.
+
+|Value type|Scalar class|Notes|
+|-|-|-|
+|fixed-width|`fixed_width_scalar<T>`| `T` can be any fixed-width type|
+|numeric|`numeric_scalar<T>` | `T` can be `int8_t`, `int16_t`, `int32_t`, `int_64_t`, `float` or `double`|
+|fixed-point|`fixed_point_scalar<T> | `T` can be `cudf::numeric::decimal32` or `cudf::numeric::decimal64`|
+|timestamp|`timestamp_scalar<T>` | `T` can be `timestamp_D`, `timestamp_s`, etc.|
+|duration|`duration_scalar<T>` | `T` can be `duration_D`, `duration_s`, etc.|
+|string|`string_scalar`| This class object is immutable|
+
+### Construction
+`scalar`s can be created using either their respective constructors or using factory functions like 
+`make_numeric_scalar()`, `make_timestamp_scalar()` or `make_string_scalar()`. 
+
+### Casting
+All the factory methods return a `unique_ptr<scalar>` which needs to be statically downcasted to 
+its respective scalar class type before accessing its value. Their validity (nullness) can be 
+accessed without casting. Generally, the value needs to be accessed from a function that is aware 
+of the value type e.g. a functor that is dispatched from `type_dispatcher`. To cast to the 
+requisite scalar class type given the value type, use the mapping utility `scalar_type_t` provided 
+in `type_dispatcher.hpp` : 
+
+```c++
+//unique_ptr<scalar> s = make_numeric_scalar(...);
+
+using ScalarType = cudf::scalar_type_t<T>;
+// ScalarType is now numeric_scalar<T>
+auto s1 = static_cast<ScalarType *>(s.get());
+```
+
+### Passing to device
+Each scalar type has a corresponding non-owning device view class which allows access to the value 
+and its validity from the device. This can be obtained using the function 
+`get_scalar_device_view(ScalarType s)`. Note that a device view is not provided for a base scalar 
+object, only for the derived typed scalar class objects.
+
+# libcudf++ API and Implementation
+
+## Streams
+
+CUDA streams are not yet exposed in external libcudf APIs. However, in order to ease the transition 
+to future use of streams, all libcudf APIs that allocate device memory or execute a kernel should be 
+implemented using asynchronous APIs on the default stream (e.g., stream 0).
+
+The recommended pattern for doing this is to make the definition of the external API invoke an 
+internal API in the `detail` namespace. The internal `detail` API has the same parameters as the 
+public API, plus a `rmm::cuda_stream_view` parameter at the end defaulted to 
+`rmm::cuda_stream_default`. The implementation should be wholly contained in the `detail` API 
+definition and use only asynchronous versions of CUDA APIs with the stream parameter.
+
+In order to make the `detail` API callable from other libcudf functions, it should be exposed in a 
+header placed in the `cudf/cpp/include/detail/` directory.
```

Comment:
Seems we should mention this in the Directory Structure section as well?

Reply:
It's mentioned in the section on `detail` namespace, but added a sentence about where these detail headers should live to the Directory Structure section.